rule_builder.py

```diff
@@ -3,20 +3,112 @@
+            raise ValueError(f"Rule {name} not found")
+        return custom_rule_classes.get(name) or DEFAULT_RULES[name]
+
+    @classmethod
+    def rule_from_json(cls, data: Mapping[str, Any]) -> "Rule[Self]":
+        name = data.get("rule", "")
+        rule_class = cls.get_rule_cls(name)
+        return rule_class.from_json(data)
+
+    def resolve_rule(self, rule: "Rule[Self]") -> "Rule.Resolved":
+        resolved_rule = rule.resolve(self)
+        if resolved_rule.cacheable:
+            for item_name, rule_ids in resolved_rule.item_dependencies().items():
+                self.rule_dependencies[item_name] |= rule_ids
+        return resolved_rule
+
+    def register_rule_connections(self, resolved_rule: "Rule.Resolved", entrance: "Entrance") -> None:
+        for indirect_region in resolved_rule.indirect_regions():
+            self.multiworld.register_indirect_condition(self.get_region(indirect_region), entrance)
+
+    def set_rule(self, spot: "Location | Entrance", rule: "Rule[Self]") -> None:
+        resolved_rule = self.resolve_rule(rule)
+        spot.access_rule = resolved_rule.test
+        if self.explicit_indirect_conditions and isinstance(spot, Entrance):
+            self.register_rule_connections(resolved_rule, spot)
+
+    def create_entrance(
+        self,
+        from_region: "Region",
+        to_region: "Region",
+        rule: "Rule[Self] | None",
+    ) -> "Entrance | None":
+        resolved_rule = None
+        if rule is not None:
+            resolved_rule = self.resolve_rule(rule)
+            if resolved_rule.always_false:
+                return None
+
+        entrance = from_region.connect(to_region, rule=resolved_rule.test if resolved_rule else None)
+        if resolved_rule is not None:
+            self.register_rule_connections(resolved_rule, entrance)
+        return entrance
+
+    @override
+    def collect(self, state: "CollectionState", item: "Item") -> bool:
+        changed = super().collect(state, item)
+        if changed and getattr(self, "rule_dependencies", None):
+            player_results: dict[int, bool] = state.rule_cache[self.player]
+            for rule_id in self.rule_dependencies[item.name]:
+                _ = player_results.pop(rule_id, None)
+        return changed
+
+    @override
+    def remove(self, state: "CollectionState", item: "Item") -> bool:
+        changed = super().remove(state, item)
+        if changed and getattr(self, "rule_dependencies", None):
+            player_results: dict[int, bool] = state.rule_cache[self.player]
+            for rule_id in self.rule_dependencies[item.name]:
+                _ = player_results.pop(rule_id, None)
+        return changed
+
+
+TWorld = TypeVar("TWorld", bound=RuleWorldMixin, contravariant=True)  # noqa: PLC0105
+
+
+@dataclass_transform()
+def custom_rule(world_cls: "type[TWorld]", init: bool = True) -> "Callable[..., type[Rule[TWorld]]]":
+    def decorator(rule_cls: "type[Rule[TWorld]]") -> "type[Rule[TWorld]]":
```

Comment:
Do you need functools.wraps to prevent metadata loss, or does passing it via dataclass cover it

Reply:
This and `dataclass` return the class directly rather than a wrapper so nothing needs to be wrapped